org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/FileEventHandler.java

```diff
@@ -129,49 +149,129 @@ private static WorkspaceEdit computePackageRenameEdit(FileRenameEvent[] renameEv
-					} finally {
-						renameMonitor.done();
+			IPackageFragment oldPackageFragment = resolvePackageFragment(oldLocation, sourcePaths);
+			SubMonitor renameMonitor = submonitor.split(100);
+			try {
+				if (oldPackageFragment != null && !oldPackageFragment.isDefaultPackage() && oldPackageFragment.getResource() != null) {
+					String oldPackageName = oldPackageFragment.getElementName();
+					int lastDot = oldPackageName.lastIndexOf(".");
+					String newPackageName = lastDot < 0 ? newLocation.lastSegment() :
+						oldPackageName.subSequence(0, lastDot + 1) + newLocation.lastSegment();
+					oldPackageFragment.getResource().refreshLocal(IResource.DEPTH_INFINITE, null);
+					if (oldPackageFragment.exists()) {
+						ResourcesPlugin.getWorkspace().run((pm) -> {
+							WorkspaceEdit edit = getRenameEdit(oldPackageFragment, newPackageName, pm);
+							root[0] = ChangeUtil.mergeChanges(root[0], edit, true);
+						}, oldPackageFragment.getSchedulingRule(), IResource.NONE, renameMonitor);
 					}
-
-					break;
 				}
+			} catch (CoreException e) {
+				JavaLanguageServerPlugin.logException("Failed to compute the package rename update", e);
+			} finally {
+				renameMonitor.done();
 			}
 		}
 
 		submonitor.done();
 		return ChangeUtil.hasChanges(root[0]) ? root[0] : null;
 	}
 
+	private static WorkspaceEdit computeMoveEdit(FileRenameEvent[] moveEvents, SourcePath[] sourcePaths, IProgressMonitor monitor) {
+		IPath[] newPaths = Stream.of(moveEvents).map(event -> ResourceUtils.filePathFromURI(event.newUri)).toArray(IPath[]::new);
+		IPath destinationPath = getLongestCommonPath(newPaths);
+		if (destinationPath == null) {
+			return null;
+		}
+
+		IPackageFragment destinationPackage = resolvePackageFragment(destinationPath, sourcePaths);
+		if (destinationPackage == null) {
+			return null;
+		}
+
+		List<IJavaElement> javaElements = new ArrayList<>();
+		IPath[] oldPaths = Stream.of(moveEvents).map(event -> ResourceUtils.filePathFromURI(event.oldUri)).toArray(IPath[]::new);
+		for (int i = 0; i < oldPaths.length; i++) {
+			if (oldPaths[i].toFile().isFile()) {
+				ICompilationUnit cu = JDTUtils.resolveCompilationUnit(moveEvents[i].oldUri);
+				if (cu != null && cu.getJavaProject() != null && cu.getJavaProject().isOnClasspath(cu)) {
+					javaElements.add(cu);
+				}
+			}
+		}
+
+		if (javaElements.isEmpty()) {
+			return null;
+		}
+
+		IReorgDestination packageDestination = ReorgDestinationFactory.createDestination(destinationPackage);
+		WorkspaceEdit[] root = new WorkspaceEdit[1];
+		try {
+			ResourcesPlugin.getWorkspace().run((pm) -> {
+				root[0] = MoveHandler.move(new IResource[0], javaElements.toArray(new IJavaElement[0]), packageDestination, true, pm);
+			}, monitor);
+		} catch (CoreException e) {
+			JavaLanguageServerPlugin.logException("Failed to compute the move update", e);
+		}
+
+		return ChangeUtil.hasChanges(root[0]) ? root[0] : null;
+	}
+
+	private static IPath getLongestCommonPath(IPath[] paths) {
```

Comment:
looks like something that belongs to a utility class, that could even deserve specific unit tests

Reply:
done. Moved to ResourceUtils, and added unit test.